### PR TITLE
new package: MOLGW

### PR DESCRIPTION
**Pull request by bruneval**
Create the spack packaging for [MOLGW](https://github.com/bruneval/molgw), a many-body perturbation theory code in Gaussian basis

### CODE REVIEW

**Comment by bruneval** [comment]
@tgamblin sorry for mentioning you personally.
But it looks like my pull request has been forgotten for long.

It is very simple: it simply adds a new package file, it leaves all the other packages untouched, and it passes all the CI tests.

I think this is a straightforward addition.

Thanks for your time.
Fabien

File: var/spack/repos/builtin/packages/molgw/package.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        # Write configuration file
+        with open("my_machine.arch", "w") as f:
+            for k, v in flags.items():
+                f.write(k + "=" + v + "\n")
+
+    def build(self, spec, prefix):
+        make()
+
+    def install(self, spec, prefix):
+        make("install")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/molgw/package.py]: This shouldn't be needed (it's inherited by the base package)

File: var/spack/repos/builtin/packages/molgw/package.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    version("3.2", sha256="a3f9a99db52d95ce03bc3636b5999e6d92b503ec2f4afca33d030480c3e10242")
+
+    depends_on("blas")
+    depends_on("lapack")
+    depends_on("libxc@5:")
+    depends_on("libcint+pypzpx+coulomb_erf")
+
+    variant("openmp", default=False, description="Build with OpenMP support")
+
+    variant("scalapack", default=False, description="Build with ScaLAPACK support")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/molgw/package.py]: (minor) Can we move variants above all the `depends_on` directives?

File: var/spack/repos/builtin/packages/molgw/package.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    # enforce scalapack capable mkl when asking +scalapack (and using intel-oneapi-mkl)
+    depends_on("intel-oneapi-mkl+cluster", when="+scalapack ^intel-oneapi-mkl")
+    # enforce threaded openblas when asking +openmp (and using openblas)
+    depends_on("openblas threads=openmp", when="+openmp ^openblas")
+
+    # variant('hdf5', default=False, description='Build with HDF5 support')
+    # depends_on('hdf5', when='+hdf5')
+
+    def _get_mkl_ld_flags(self, spec):
+        mklroot = str(getenv("MKLROOT"))
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/molgw/package.py:46]: Can't we use `spec["blas"].libs` etc. to abstract the provider?

File: var/spack/repos/builtin/packages/molgw/package.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            command.extend(["-c", "intel_f"])
+            if "+openmp" in spec:
+                command.extend(["-o", "tbb"])
+        elif "%gcc" in spec:
+            command.extend(["-c", "gnu_f"])
+            if "+openmp" in spec:
+                command.extend(["-o", "gomp"])
+
+        if "+scalapack" in spec:
+            command.extend(["--cluster_library=scalapack"])
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/molgw/package.py:64]: Similarly for `scalapack`

**Review by alalazo — CHANGES_REQUESTED** [review]
I left a few comments. Let me know what you think.

File: var/spack/repos/builtin/packages/molgw/package.py (hunk 1 above)
Comment (bruneval) [reply var/spack/repos/builtin/packages/molgw/package.py]: Fixed!

File: var/spack/repos/builtin/packages/molgw/package.py (hunk 2 above)
Comment (bruneval) [reply var/spack/repos/builtin/packages/molgw/package.py]: Fixed!

File: var/spack/repos/builtin/packages/molgw/package.py (hunk 4 above)
Comment (bruneval) [reply var/spack/repos/builtin/packages/molgw/package.py:64]: Dear Alalazo,

I completely agree with you in principle.

The only problem is that the spec["scalapack"].libs is not correctly set in the case of `intel-oneapi-mkl`.

Indeed, in `intel-oneapi-mkl`, the variant `+cluster` adds the linking to intelmpi only.

intel-oneapi-mkl/package.py:
```
        if "+cluster" in self.spec:
            libs.append(self._xlp64_lib("libmkl_blacs_intelmpi"))
```
This linking doesn't work with openmpi for instance.

One should fix intel-oneapi-mkl first. But I'm afraid to touch such an important package by myself.
I reported this problem in the issue #37459 one month ago...

Best,
Fabien

File: var/spack/repos/builtin/packages/molgw/package.py (hunk 3 above)
Comment (bruneval) [reply var/spack/repos/builtin/packages/molgw/package.py:46]: See my next comment.

**Comment by bruneval** [comment]
Dear @alalazo 
I think I addressed all your comments.

Please let me know if something is still missing.
Best,
Fabien

File: var/spack/repos/builtin/packages/molgw/package.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            if "openmpi" in spec:
+                command.extend(["-m", "openmpi"])
+            elif "mpich" in spec:
+                command.extend(["-m", "mpich2"])
+            elif "intelmpi" in spec:
+                command.extend(["-m", "intelmpi"])
+        # only works with recent python3
+        # result = run(command,capture_output=True, text=True)
+        # return result.stdout.strip()
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/molgw/package.py]:
```suggestion
```

File: var/spack/repos/builtin/packages/molgw/package.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        result = run(command, stdout=PIPE)
+        # with open("my_machine.arch", "w") as f:
+        #    f.write("#" + " ".join(command) + "\n")
+        #    f.write("#" + result.stdout.decode(encoding="utf-8").strip() + "\n")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/molgw/package.py]:
```suggestion
```

File: var/spack/repos/builtin/packages/molgw/package.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+    depends_on("scalapack", when="+scalapack")
+    depends_on("mpi", when="+scalapack")
+
+    # enforce scalapack-capable mkl when asking +scalapack (and using intel-oneapi-mkl)
+    depends_on("intel-oneapi-mkl+cluster", when="+scalapack ^intel-oneapi-mkl")
+    # enforce threaded openblas when asking +openmp (and using openblas)
+    depends_on("openblas threads=openmp", when="+openmp ^openblas")
+
+    def _get_mkl_ld_flags(self, spec):
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/molgw/package.py:45]: This method seems to belong more to some `Intel*` base class, than here. @rscohn2 Do you know if we have anything similar to this?

File: var/spack/repos/builtin/packages/molgw/package.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    MOLGW employs standard Gaussian basis set.
+    """
+
+    homepage = "https://github.com/bruneval/molgw"
+    url = "https://github.com/bruneval/molgw/archive/v3.2.tar.gz"
+    git = "https://github.com/bruneval/molgw.git"
+
+    maintainers("bruneval")
+
+    # version("rolling-release", branch="master")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/molgw/package.py]:
```suggestion
```

File: var/spack/repos/builtin/packages/molgw/package.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    version("3.2", sha256="a3f9a99db52d95ce03bc3636b5999e6d92b503ec2f4afca33d030480c3e10242")
+
+    # variants
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/molgw/package.py]:
```suggestion
```

File: var/spack/repos/builtin/packages/molgw/package.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    variant("openmp", default=False, description="Build with OpenMP support")
+    variant("scalapack", default=False, description="Build with ScaLAPACK support")
+
+    # dependences
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/molgw/package.py]:
```suggestion
```

**Review by alalazo — CHANGES_REQUESTED** [review]
A question, mainly for @rscohn2 and a minor request to remove commented out code and redundant comments.

File: var/spack/repos/builtin/packages/molgw/package.py (hunk 7 above)
Comment (rscohn2) [reply var/spack/repos/builtin/packages/molgw/package.py:45]: If it depends on `intel-oneapi-mkl +cluster threads=openmp` or `intel-mkl threads=openmp ^mpi` then `spec["lapack"].libs` should have everything you need.

Comment (bruneval) [reply var/spack/repos/builtin/packages/molgw/package.py:45]: As I wrote in issue #37459, ```intel-oneapi-mkl +cluster threads=openmp``` only works when linking against `intel-mpi`.

See the corresponding lines in `intel-oneapi-mkl/package.py`:
```
        if "+cluster" in self.spec:
            libs.append(self._xlp64_lib("libmkl_blacs_intelmpi"))
```

My coding allows the linking against openmpi.

I agree that in principle the `intel-oneapi-mkl` should be fixed first, but the issue I raised remained unanswered.

Comment (bruneval) [reply var/spack/repos/builtin/packages/molgw/package.py:45]: If you prefer, I can propose a pull request to `intel-oneapi-mkl`. But this package is quite central and I would need some back-up.

**Review by alalazo — APPROVED** [review]
Le'ts get this recipe in and fix the intel logic in a following PR

**Pull request merged**